84-sarcastic-fringehead/technical_specification.md

```diff
@@ -0,0 +1,238 @@
+
+If a checksum discrepancy is found, the DHFS rejects the upload and posts a `FileUploadReport` to the FIS's HTTP API which indicates that the file did not pass inspection. If checksums match and there are no other errors during upload, the DHFS accepts the upload and the `FileUploadReport` sent to the FIS reflects that the file passed inspection.
+
+In both cases, the `FileUploadReport` is broadcasted by the FIS as a Kafka event and picked up by the UCS, which would update its local `FileUpload` accordingly. If the upload is successful, the UCS deletes the file from the `inbox`.
+
+#### Interrogation Process in List Format
+- [Per File]
+  - Reads the first file part to sift for the Crypt4GH envelope
+  - Decrypts the envelope using the configured private key (specific to the Data Hub)
+  - Obtains the original file secret for decryption
+  - Generates a new file secret for re-encryption
+  - Sends the new file secret to the FIS's HTTP API
+    - This is done right away instead of waiting to see if re-encryption is successful 
+  - Calculates the content starting position (offset) from the envelope length
+  - Initiates a multipart upload with the Data Hub's `interrogation` bucket
+  - Streams the object from the Data Hub's `inbox` bucket chunk-by-chunk
+  - [Per File Chunk]
+    - Decrypts the chunk
+    - Updates the SHA-256 checksum over the unencrypted content
+    - Re-encrypts the chunk using the newly generated file secret
+    - Updates the MD5 checksum over the encrypted file content
+    - Uploads the re-encrypted chunk to the `interrogation` bucket
+  - Compares the unencrypted file's checksum against the one reported by the submitter during upload, and the encrypted checksum against the one calculated by S3
+  - Sends a `FileUploadReport` to the FIS's HTTP API
+
+The final duty of the DHFS is to clean up files from the `interrogation` bucket after archival. It does this by polling the FIS for a list of files that should be deleted. The FIS sources this list from the `FileInternallyRegistered` events published by the IFRS. When the DHFS deletes a file, it contacts the FIS HTTP API with the deleted file ID. This mechanism removes the need for the DHFS to keep track of the files in its `interrogation` bucket, provides idempotent behavior, and lets the DHFS carry out its own deletions (as opposed to giving the IFRS this duty). If the DHFS fails to find a file during deletion, it logs the absence but otherwise treats it as a successful deletion.
+
```

Comment:
I think this would be easier to read and clearer to understand if the FIS and DHFS descriptions swap places and try to be as centered as can be on their own descriptions.